Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+    case RoundingMode::TRUNK:
+        return static_cast<T>(truncl(value));
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+template<Arithmetic T>
+ALWAYS_INLINE T SoftFPU::fpu_round_checked(long double value)
+{
+    T result = fpu_round<T>(value);
+    if (auto rnd = value - result) {
+        if (rnd > 0)
+            set_c1(1);
+        else
+            set_c1(0);
+        fpu_set_exception(FPU_Exception::Precision);
+    }
+    return result;
+}
+
+template<FloatingPoint T>
+ALWAYS_INLINE T SoftFPU::fpu_convert(long double value) const
+{
+    // FIXME: actually round the right way
+    return static_cast<T>(value);
+}
+template<FloatingPoint T>
+ALWAYS_INLINE T SoftFPU::fpu_convert_checked(long double value)
+{
+    T result = fpu_convert<T>(value);
+    if (auto rnd = value - result) {
+        if (rnd > 0)
+            set_c1(1);
+        else
+            set_c1(0);
+        fpu_set_exception(FPU_Exception::Precision);
+    }
+    return result;
+}
+
+// Instructions
+
+// DATA TRANSFER
+void SoftFPU::FLD_RM32(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_push(fpu_get(insn.modrm().register_index()));
+    } else {
+        auto new_f32 = insn.modrm().read32(m_cpu, insn);
+        // FIXME: Respect shadow values
+        fpu_push(bit_cast<float>(new_f32.value()));
+    }
+}
+void SoftFPU::FLD_RM64(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto new_f64 = insn.modrm().read64(m_cpu, insn);
+    // FIXME: Respect shadow values
+    fpu_push(bit_cast<double>(new_f64.value()));
+}
+void SoftFPU::FLD_RM80(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+
+    // long doubles can be up to 128 bits wide in memory for reasons (alignment) and only uses 80 bits of precision
+    // GCC uses 12 bytes in 32 bit and 16 bytes in 64 bit mode
+    // so in the 32 bit case we read a bit to much, but that shouldn't be an issue.
+    // FIXME: Respect shadow values
+    u128 new_f80 = insn.modrm().read128(m_cpu, insn).value();
+
+    fpu_push(*(long double*)new_f80.bytes().data());
+}
+
+void SoftFPU::FST_RM32(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    float f32 = fpu_convert_checked<float>(fpu_get(0));
+
+    if (fpu_is_set(0))
+        insn.modrm().write32(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u32>(f32)));
+    else
+        insn.modrm().write32(m_cpu, insn, ValueWithShadow<u32>(bit_cast<u32>(f32), 0u));
+}
+void SoftFPU::FST_RM64(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(insn.modrm().register_index(), fpu_get(0));
+    } else {
+        double f64 = fpu_convert_checked<double>(fpu_get(0));
+        if (fpu_is_set(0))
+            insn.modrm().write64(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u64>(f64)));
+        else
+            insn.modrm().write64(m_cpu, insn, ValueWithShadow<u64>(bit_cast<u64>(f64), 0ULL));
+    }
+}
+
```

Comment:
no fst_RM80?

Reply:
yeah that one always pops
https://www.felixcloutier.com/x86/fst:fstp